src/CommandLineUtils/Abstractions/DefaultValueParserFactory.cs

```diff
@@ -0,0 +1,123 @@
+﻿namespace McMaster.Extensions.CommandLineUtils.Abstractions
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.Text;
+
+    /// <summary>
+    /// A factory creating generic implementations of <see cref="IValueParser{T}"/>. The implementations are based
+    /// on automatically located <see cref="TypeConverter"/> classes that are suitable for parsing.
+    /// </summary>
+    sealed class DefaultValueParserFactory
+    {
+        const int DefaultMaxCacheCapacity = 100;
+
+        #region Private Fields
+
+        private readonly int _maxCacheCapacity;
+
+        [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
+        readonly Dictionary<Type, IValueParser> _parsersByTargetType = new Dictionary<Type, IValueParser>();
+
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        readonly HashSet<Type> _notSupportedTypes = new HashSet<Type>();
+
+        #endregion
+
+        public DefaultValueParserFactory(int maxCacheCapacity = DefaultMaxCacheCapacity)
+        {
+            _maxCacheCapacity = maxCacheCapacity;
+            if (_maxCacheCapacity < 1) throw new ArgumentOutOfRangeException(nameof(maxCacheCapacity));
+        }
+
+        [DebuggerStepThrough]
+        public bool TryGetParser<T>(out IValueParser<T> parser)
+        {
+            if (TryGetParser<T>(out IValueParser generalizedParser))
+            {
+                parser = (IValueParser<T>)generalizedParser;
+                return true;
+            }
+
+            parser = null;
+            return false;
+        }
+
+        public bool TryGetParser<T>(out IValueParser parser)
+        {
+            var targetType = typeof(T);
+            if (_notSupportedTypes.Contains(targetType))
+            {
+                parser = null;
+                return false;
+            }
+
+            if (_parsersByTargetType.TryGetValue(targetType, out parser))
+            {
+                Debug.Assert(targetType == parser.TargetType);
+                return true;
+            }
+
+            var converter = TypeDescriptor.GetConverter(targetType);
```

Comment:
I've never used this API before. Do you know what other kinds of types this supports? Also, does this incidentally implement support for TypeConverterAttribute? If so, this is a two-birds-one-stone PR and also resolves #62.

Reply:
Indeed, the change made does address #62. The requested functionality is implemented in the form of implicit fallback, when no custom value parser is registered for an argument type. In principle the change introduced support for a wide range of CLR types (e.g. TimeSpan, Version etc.). And yes, your right. The TypeConverterAttribute functionality is enabled with this change. So custom types annotated with TypeConverterAttribute will be parsed automatically.